File: 540/main.go
package main

import (
	"fmt"
)

func main() {
	fmt.Println(singleNonDuplicate([]int{1,1,2,2,3,4,4,5,5,6,6}))
	fmt.Println(singleNonDuplicate([]int{1,1,2,2,3,3,4,4,5,5,6,6,7}))
	fmt.Println(singleNonDuplicate([]int{1,1,2,2,3,3,4,4,5,5,6}))
	fmt.Println(singleNonDuplicate([]int{1,2,2,3,3,4,4,5,5,6}))
	fmt.Println(singleNonDuplicate([]int{1,1,2,3,3}))
}

func singleNonDuplicate(nums []int) int {
	if len(nums) == 1 {
		return nums[0]
	}
	n := len(nums)-1
	for i := 0; i <= len(nums) / 2; i+=2 {
		if nums[i] != nums[i+1] {
			return nums[i]
		}
		if nums[n-i] != nums[n-i-1] {
			return nums[n-i]
		}
	}
	return 0
}

File: 540/main.rs
struct Solution {}

impl Solution {
    pub fn single_non_duplicate(nums: Vec<i32>) -> i32 {
        let mut low = 0;
        let mut high = nums.len() - 1;
        while low < high {
            let mid = low + (high-low) / 2;
            if nums[mid] == nums[mid^1] {
                low = mid + 1;
            } else {
                high = mid;
            }
        }
        nums[low]
    }
}

fn main() {
    println!("{}", Solution::single_non_duplicate(vec![1,1,2,3,3]));
}